upcloud/request/managed_database.go

```diff
@@ -0,0 +1,549 @@
+	qv := url.Values{}
+	if g.Period != "" {
+		qv.Set("period", g.Period)
+	}
+	return fmt.Sprintf("/database/%s/metrics?%s", g.UUID, qv.Encode())
+}
+
+// GetManagedDatabaseLogsRequest represents a request to get managed database instance logs
+type GetManagedDatabaseLogsRequest struct {
+	// UUID selects a managed database instance to query logs from
+	UUID string
+	// Order sets the log sort order. See:
+	// 	upcloud.ManagedDatabaseLogOrderAscending
+	// 	upcloud.ManagedDatabaseLogOrderDescending
+	Limit  int
+	Offset string
+	Order  string
+}
+
+// RequestURL implements the Request interface
+func (g *GetManagedDatabaseLogsRequest) RequestURL() string {
+	qv := url.Values{}
+	if g.Limit != 0 {
+		qv.Set("limit", strconv.Itoa(g.Limit))
+	}
+	if g.Offset != "" {
+		qv.Set("offset", g.Offset)
+	}
+	if g.Order != "" {
+		qv.Set("order", g.Order)
+	}
+	if g.Offset == "" && g.Order == upcloud.ManagedDatabaseLogOrderAscending {
+		qv.Set("offset", "0")
+	}
+	return fmt.Sprintf("/database/%s/logs?%s", g.UUID, qv.Encode())
+}
+
+type GetManagedDatabaseQueryStatisticsRequest struct {
+	// UUID selects a managed database instance to query statistics from
+	UUID string
+	// Limit sets the upper bound how many query stats to fetch
+	Limit int
+	// Offset skips n query stat rows before returning them. It can be used to iteratively fetch all.
+	Offset int
+}
+
+// RequestURL implements the Request interface
+func (g *GetManagedDatabaseQueryStatisticsRequest) RequestURL() string {
+	qv := url.Values{}
+	if g.Limit != 0 {
+		qv.Set("limit", strconv.Itoa(g.Limit))
+	}
+	if g.Offset != 0 {
+		qv.Set("offset", strconv.Itoa(g.Offset))
+	}
+	return fmt.Sprintf("/database/%s/query-statistics?%s", g.UUID, qv.Encode())
+}
+
+// ManagedDatabaseMaintenanceRequest represents the set time of week when automatic maintenance operations are allowed
+type ManagedDatabaseMaintenanceRequest struct {
+	DayOfWeek string `json:"dow,omitempty"`
+	Time      string `json:"time,omitempty"`
+}
+
+// ManagedDatabasePropertiesRequest is a Properties helper type for CreateManagedDatabaseRequest and ModifyManagedDatabaseRequest.
+// It allows initialisation by chaining the Set methods.
+//
+// For example:
+// 	req := CreateManagedDatabaseRequest{}; req.Properties.SetString("foo", "bar").Set("test", customValue)
+type ManagedDatabasePropertiesRequest map[string]interface{}
+
+// Set associates key with a any type of value. The underlying map is initialised if it's nil
+func (m *ManagedDatabasePropertiesRequest) Set(name string, value interface{}) *ManagedDatabasePropertiesRequest {
+	if m == nil {
+		return nil
+	}
+	if *m == nil {
+		*m = make(ManagedDatabasePropertiesRequest)
+	}
+	(*m)[name] = value
+	return m
+}
+
+// SetString associates key with a string value. The underlying map is initialised if it's nil
+func (m *ManagedDatabasePropertiesRequest) SetString(name string, value string) *ManagedDatabasePropertiesRequest {
```

Comment:
I'm guessing these are just for making the consumer code more self-documenting? Should we make `Set` private in that case, to enforce their usage?

Reply:
yes, though you can always just mutate the map. There are also some properties that have inner map as values. We provide setters for well-known properties that are common to all engines.